srcjs/input_binding_fileinput.js

```diff
@@ -266,11 +293,169 @@ $.extend(fileInputBinding, {
+      }, 0);
+    });
+  },
+  disableDraghover: function($el) {
+    $el.off(".dragHover");
+  },
+  enableDocumentEvents: function() {
+    let $doc = $(document);
+
+    this.enableDraghover($doc);
+    $doc.on("draghoverstart.fileDrag", e => {
+      $fileInputs.trigger("showZone.fileDrag");
+    });
+    $doc.on("draghoverend.fileDrag", e => {
+      $fileInputs.trigger("hideZone.fileDrag");
+    });
+    // Enable the drop event and prevent download if the file is dropped outside
+    // of a drop zone.
+    $doc.on("dragover.fileDrag drop.fileDrag", e => {
+      e.preventDefault();
+    });
+  },
+  disableDocumentEvents: function() {
+    let $doc = $(document);
+
+    $doc.off(".fileDrag");
+    this.disableDraghover($doc);
+  },
+  zoneStyles: {
+    "plain": {
+      "box-shadow": "none"
+    },
+    "activated": {
+      "box-shadow": "inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(102, 175, 233, .6)"
+    },
+    "over": {
+      "box-shadow": "inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(76, 174, 76, .6)"
+    }
+  },
+  zoneEvents: [
+    "showZone.fileDrag",
+    "hideZone.fileDrag",
+    "dragenter",
+    "dragleave",
+    "drop"
+  ].join(" "),
+  canSetFiles: function(fileList) {
+    var testEl = document.createElement("input");
+    testEl.type = "file";
+    try {
+      testEl.files = fileList;
+    } catch (e) {
+      return false;
+    }
+    return true;
+  },
+  handleDrop: function(e, el) {
+    const files = e.originalEvent.dataTransfer.files,
+          $el   = $(el);
+    if (files === undefined || files === null) {
+      // 1. The FileList object isn't supported by this browser, and
+      // there's nothing else we can try. (< IE 10)
+      console.log("Dropping files is not supported on this browser. (no FileList)");
+    } else if (!this.canSetFiles(files)) {
+      // 2. The browser doesn't support assigning a type=file input's .files
+      // property, but we do have a FileList to work with. (IE10+)
+      $el.val("");
+      uploadDroppedFilesIE10Plus(el, files);
+      window.setTimeout(() => {
+        // We manually trigger the hideZone event because for some reason the
+        // document-level handler is disabled by the drop event.
+        $fileInputs.trigger("hideZone.fileDrag");
+        // Because the document handlers are removed inexplicably on IE10, we
+        // re-add them here.
+        this.disableDocumentEvents();
+        this.enableDocumentEvents();
+      }, 0);
+    } else {
+      // 3. The browser supports FileList and input.files assignment.
+      // (Chrome, Safari)
+      $el.val("");
+      el.files = e.originalEvent.dataTransfer.files;
+    }
+  },
   subscribe: function(el, callback) {
-    $(el).on('change.fileInputBinding', uploadFiles);
+    let $el        = $(el),
+        $zone      = this.getZone(el),
+        getState   = () => $el.data("state"),
+        setState   = (newState) => $el.data("state", newState),
+        transition = multimethod()
```

Comment:
I think it would be helpful to have a short description of the state machine, and how the JS events and state transitions are wired together.